src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs

```diff
@@ -15952,22 +15985,367 @@ public static Vector512<T> Invoke(Vector512<T> x)
+            //
+            // 3. Reconstruction
+            //      Hence, cos(x) = sin(x + pi/2) = (-1)^N * sin(f)
+
+            public static bool Vectorizable => typeof(T) == typeof(float) || typeof(T) == typeof(double);
+
             public static T Invoke(T x) => T.Cos(x);
-            public static Vector128<T> Invoke(Vector128<T> x) => throw new NotSupportedException();
-            public static Vector256<T> Invoke(Vector256<T> x) => throw new NotSupportedException();
-            public static Vector512<T> Invoke(Vector512<T> x) => throw new NotSupportedException();
+
+            public static Vector128<T> Invoke(Vector128<T> x)
+            {
+                if (typeof(T) == typeof(float))
+                {
+                    return CosOperatorSingle.Invoke(x.AsSingle()).As<float, T>();
+                }
+                else
+                {
+                    Debug.Assert(typeof(T) == typeof(double));
+                    return CosOperatorDouble.Invoke(x.AsDouble()).As<double, T>();
+                }
+            }
+
+            public static Vector256<T> Invoke(Vector256<T> x)
+            {
+                if (typeof(T) == typeof(float))
+                {
+                    return CosOperatorSingle.Invoke(x.AsSingle()).As<float, T>();
+                }
+                else
+                {
+                    Debug.Assert(typeof(T) == typeof(double));
+                    return CosOperatorDouble.Invoke(x.AsDouble()).As<double, T>();
+                }
+            }
+
+            public static Vector512<T> Invoke(Vector512<T> x)
+            {
+                if (typeof(T) == typeof(float))
+                {
+                    return CosOperatorSingle.Invoke(x.AsSingle()).As<float, T>();
+                }
+                else
+                {
+                    Debug.Assert(typeof(T) == typeof(double));
+                    return CosOperatorDouble.Invoke(x.AsDouble()).As<double, T>();
+                }
+            }
+        }
+
+        /// <summary>float.Cos(x)</summary>
+        private readonly struct CosOperatorSingle : IUnaryOperator<float, float>
+        {
+            internal const uint MaxVectorizedValue = 0x4A989680u;
+            internal const uint SignMask = 0x7FFFFFFFu;
+            private const float AlmHuge = 1.2582912e7f;
+            private const float Pi_Tail1 = 8.742278e-8f;
+            private const float Pi_Tail2 = 3.430249e-15f;
+            private const float C1 = -0.16666657f;
+            private const float C2 = 0.008332962f;
+            private const float C3 = -1.9801206e-4f;
+            private const float C4 = 2.5867037e-6f;
+
+            public static bool Vectorizable => true;
+
+            public static float Invoke(float x) => float.Cos(x);
+
+            public static Vector128<float> Invoke(Vector128<float> x)
+            {
+                Vector128<uint> uxMasked = x.AsUInt32() & Vector128.Create(SignMask);
```

Comment:
This one should be cheaper as `Vector128.Abs(x).AsUInt32()`, that's all it's doing